inspirehep/ext/cache_manager/manager.py

```diff
@@ -0,0 +1,176 @@
+            brief_search_views_to_invalidate.append(changed_reference)
+            citations_to_invalidate.append(changed_reference)
+
+        # get all references of changed_references and invalidate their citations view
+        recids_to_invalidate = []
+        query = ' or '.join(['control_number:' + str(record_id) for record_id in changed_references])
+        es_query = Query(query).search()
+        es_query.body.update({
+            "size": 9999,
+            '_source': ['references']
+        })
+        for record in es_query.records():
+            if record.get('references'):
+                for ref in record['references']:
+                    ref_recid = ref.get('recid')
+                    if ref_recid:
+                        recids_to_invalidate.append(ref_recid)
+
+        citations_to_invalidate.extend(recids_to_invalidate)
+
+        # invalidate references_view of all records that refer to changed references
+        # as the changed references have new citation count
+        query = ' or '.join(['refersto:' + str(record_id) for record_id in changed_references])
+        es_query = Query(query).search()
+        es_query.body.update({
+            "size": 9999,
+            '_source': ['control_number']
+        })
+        records_citing_changed_references = [r['control_number'] for r in es_query.records() if r.get('control_number')]
+        references_to_invalidate.extend(records_citing_changed_references)
+
+        # invalidate all
+        kwargs = {
+            self.CITATIONS: citations_to_invalidate,
+            self.CITATION: single_citations_to_invalidate,
+            self.REFERENCE: single_references_to_invalidate,
+            self.BRIEF_SEARCH: brief_search_views_to_invalidate,
+            self.REFERENCES: references_to_invalidate
+        }
+        self.invalidator.invalidate_caches(**kwargs)
+
+    def invalidate_on_changed_record(self, recid, references):
+        """Invalidates template caches in case of change of the record."""
+
+        brief_search_views_to_invalidate = []
+        references_to_invalidate = []
+        citations_to_invalidate = []
+        single_citations_to_invalidate = []
+        single_references_to_invalidate = []
+
+        # invalidate brief search view and single-citation view of changed record
+        # because they present outdated version of the changed record
+        brief_search_views_to_invalidate.append(recid)
+        single_citations_to_invalidate.append(recid)
+        single_references_to_invalidate.append(recid)
+
+        # invalidate citations views of all record's references as they contain outdated version of the record
+        for ref_recid in references:
+            citations_to_invalidate.append(ref_recid)
+
+        # get all records that refer to changed record and invalidate their references views
+        # as they contain outdated version of the record
+        es_query = Query('refersto:' + str(recid)).search()
```

Comment:
Just use plain ES APIs here (since we should substitute this with Neo4J once we attack it). In this way you can optimize a lot the query.